### DISCUSSION TITLE
Enabling redengine; ensuring running status

### DISCUSSION THREAD

**Discussion by jewettg** [Miksus/rocketry#38] [Q&A]
Just a couple of quick questions.   I have successfully dove into Python with decorators and setup Flask with Nginx and such,  and can trigger my scripts to run when the URL is visited.  That makes sense.

So how does red-engine get enabled and stay running?  Is it only running when the parent python script is running?  How would one ensure the same stability as crontab and have red-engine start up and run my tasks at start-up?

I have a server that I run all my various automation scripts, but I am currently scheduling everything via cron, and it is klunky as heck (but working) and I want to move away from this.. and red-engine sounds like it will fit the bill.   I just can figure out the answer to the above, as most of my scripts are enable and forget.  I see the results throughout my workflow (data in Splunk, emails received, and tasks completed (backups and copies), and certificates renewed.  

To add.. I have never seen a python script that follows the phrase "terminate and stay ready", old school from the days of assembly and writing machine code. (Like Linux processes, Windows Services, etc..) Like I said, trigger scripts to run via decorators and utilizing nginx to stay running in the background that accepts the connections and hosts the flask application (lots of configuration).

This would be new to me, new paradigm for me.
Thanks!

**Comment by Miksus** [comment]
I probably should write an extensive documentation of the inner workings at some point but I'll try my best to describe the process.

The app (``RedEngine``) that you are probably already somewhat familiar is just an interface. Underneath there are three components related to running the system:
- RedEngine: This is just an interface actually. It's actually pretty minimal and just calls session. Meant to be the highest level of abstraction. Calls the session when starting the scheduling.
- Session: this is sort of the scheduling blueprint. This contains almost everything that defines how the scheduler works: tasks, configurations etc. This is meant to be interacted by the user also in runtime which allows you to make metatasks that modify the scheduler when it is running, pretty insane. Calls the scheduler to start the scheduling.
- Scheduler: This is the functional engine. It mostly reads the session and contains quite a lot of methods to work properly. Handles the actual scheduling.

## Scheduler
In a nutshell, Red Engine's scheduler is a simple _while_ loop that runs as long as the shut condition returns false or the shut/restart flags are not set. This _while_ loop:
- Check if shut down/restart flags are set or shut condition is reached
  - If any of those are reached, stop the while loop
  - If those are not reached:
    - Loops each task (in the order of the priority, yet an undocumented feature)
      - If task's start condition is fulfilled, then run the task
      - If task's timeout is reached, terminate the task (yet to be documented) *
      - If task's end condition is reached, terminate the task (yet to be documented) *
    - Check log queue and relog them **
    - (back to the first line)

\* Only for tasks with execution as ``thread`` or ``process``
\*\* Only for tasks with execution as ``process``

I think unparallelized scheduler (if you only used ``execution='main'``) is quite uninteresting and essentially is kind of just a fancy while loop. The interesting bit comes with the parallelization. I'll quickly go through the execution options:
- ``main`` (unparallelized, scheduler does nothing else when a task of this kind runs)
- ``thread`` (runs on separate thread)
- ``process`` (runs on separate process)

## Main
This is probably pretty straightforward to understand: 
- Log the task to run
- Execute the task (typically a Python function)
- If no errors, log success and pass the return to the session
- If errors, log failure

## Thread
This is slightly more complicated as we need to ensure the task gets logged as running correctly before its execution is checked again.

This is works as the following:
- Create a threading event ``is_running``
- Create the thread and set it running:
  - (In the new thread)
    - log running
    - Set the event ``is_running``
    - Run the task 
    - If failed, log it as failed. If successful, log it as success and set the return to the session.
  - (in the main thread where the scheduling is)
    - Wait till the ``is_running`` flag is set
    - Go back and continue scheduling

## Process
This is actually quite complex as the overhead of creating a process is quite significant (compared to thread), the objects need to be pickled and there is no shared memory.

In the scheduler, there are two extra thing related to this:
- ``log_queue``: this is a queue/pipe of log records that acts as a channel from the task processes to the main process (where the scheduler runs)
- There is a ``handle_log`` method called after each task cycle to empty the queue and log the records
  - This also handles the return values (for input-output pipelining). The log records from the queue contain also the return (if has any)

This is works as the following:
- Create a process and pass the ``log_queue`` to the task
- Start the process
  - (in between of creating the new process: pickling)
    - Some unpicklable or probably unpicklable stuff is removed. The tasks are Pydantic classes which actually are painful to operate in pickling and we need to butcher some of the attributes (which are not essential for the system) for the new process.
  - (In the new process)
    - Just remove all the handlers in the logger. We are in separate memory
    - Add a modified queue handler and attach the `log_queue`` to it
    - log the task running
    - Run the task 
    - If failed, log failed. If successful, log successful and attach the output value to the log record
  - (in the main process)
    - Wait till there is a running log coming from the ``log_queue`` or till the process dies
      - If the process died here (and did not log running), it is logged as critically failed
      - If the queue returned something else than running, it was from another task. Log that and continue waiting
      - If the queue returned running log, log that and go back and continue scheduling as usual

There are more details here:
- How to handle parameters (which may or may not be picklable)
- How to correctly attach the traceback. [Logging library's QueueHandler is not really good in this](https://github.com/python/cpython/issues/89087). This is why the QueueHandler is modified.
- How to handle the pickling


## Other nyances
- ``time.time`` is solely used for measurement of time as ``datetime.datetime.now`` is inconsistent with that and logging uses ``time.time``. Anoter curiosity I found from Python...: https://github.com/python/cpython/issues/88994
- If there is a premature error, the system crashes by default. There is a setting in the ``session.config.silence_task_prerun`` to silence it. You probably want to silence in prod but not in test.
- Only ``process`` and ``thread`` execution tasks can be terminated. Threaded tasks expect they listen an event. I'm planning on implementing it as the new-style argument but it can be already accessed with setting a parameter ``_thead_terminate_`` to your function.

## Closing words
Took me more than I thought but maybe this could act as the baseline of documenting the inner workings of the scheduler when writing those comes to the table. It actually surprises me all the details this system has.

I'm not actually sure how such things are done elsewhere but my implementation focused on ideas: 
- Switching between the execution types should be easy (for the user, was not for me)
- The main (thread and process) has always the truth and it is synchronized to maintain it. This might cause some more overhead/delay in the threading/multiprocessing but it gives robustness. The focus is more on reliability than speed. We are using Python after all.

I have successfully attached an API built with Flask on the system and partially attached a FastAPI app. The latter failed for unknown reasons related to I don't know how FastAPI works.

Lastly, did I answer the question?

**Reply by jewettg** [reply]
Thanks for the reply.  I am still trying to read it and understand it.   See my last paragraph of my question.. I am trying to figure out how I would make this tool work like cron, in that cron is available at system start/restart.  This tool is so much better and more flexible that cron.. that I would love to remove all the kludgy cron setup I have that runs my scripts at specified intervals and use your tool.

I only recently was introduced to threading and script-thread communication (celery and redis).  So I am a bit naive when it comes to understanding how to get this working in the fashion I am looking for, if possible.

**Reply by jewettg** [reply]
Right now I am using a kludgy cron execution that mixes bash shell with some script execution, stdin/stdout/stderr piping and using a third-party tool to ensure that processes do not step all over each other.
```
0 18 * * *    root    . ~/.bashrc; conda activate p3; /usr/bin/flock -w 300 /tmp/script.lck /opt/scripts/update_scripts/update_scripts.py te_test_alert_metrics >> /opt/lb-bkups/script_logs/cronlogs/cron_`date +\%Y\%m\%d`.log 2>&1; conda deactivate
```

I would obviously need to rewrite some code in my scripts to support calling them as functions (with a ```__main__``` in the file) and ensure that the entire environment is enclosed in my Python3 virtual environment (conda).
